service/s3/s3manager/copy.go

```diff
@@ -0,0 +1,471 @@
+	if c.cfg.MaxPartSize < MinUploadPartSize || c.cfg.MaxPartSize > MaxUploadPartSize {
+		return awserr.New(
+			"InvalidRequest",
+			fmt.Sprintf(
+				"expected %d <= MaxPartSize <= %d",
+				MinUploadPartSize, MaxUploadPartSize),
+			nil)
+	}
+
+	if c.cfg.MultipartCopyThreshold == 0 {
+		c.cfg.MultipartCopyThreshold = DefaultMultipartCopyThreshold
+	}
+
+	if c.cfg.MultipartCopyThreshold < MinUploadPartSize || c.cfg.MultipartCopyThreshold > MaxUploadPartSize {
+		return awserr.New(
+			"InvalidRequest",
+			fmt.Sprintf(
+				"expected %d <= MultipartCopyThreshold <= %d",
+				MinUploadPartSize, MaxUploadPartSize),
+			nil)
+	}
+
+	if c.cfg.Concurrency == 0 {
+		c.cfg.Concurrency = DefaultCopyConcurrency
+	}
+
+	if c.cfg.Concurrency <= 0 {
+		return awserr.New(
+			"InvalidRequest",
+			"expected 0 < Concurrency",
+			nil)
+	}
+
+	err := c.initSource()
+	if err != nil {
+		return err
+	}
+
+	c.src.size, err = c.getSourceSize()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// initSource derives the bucket, key, and optional versionId from the
+// CopySource input field.
+func (c *copier) initSource() error {
+	if c.in.CopySource == nil {
+		return awserr.New("InvalidRequest", "expected non-nil copy source", nil)
+	}
+
+	src, err := url.QueryUnescape(*c.in.CopySource)
+	if err != nil {
+		return awserr.New("InvalidRequest", "invalid copy source", err)
+	}
+
+	a := strings.SplitN(src, "/", 2)
+	if len(a) != 2 {
+		return awserr.New(
+			"InvalidRequest",
+			"invalid copy source: expected bucket and key",
+			nil)
+	}
+
+	b := strings.SplitN(a[1], "?versionId=", 2)
+	if len(a) > 2 {
+		return awserr.New(
+			"InvalidRequest",
+			"invalid copy source: expected at most one versionId",
+			nil)
+	}
+
+	if b[0] == "" {
+		return awserr.New(
+			"InvalidRequest",
+			"invalid copy source: expected key",
+			nil)
+	}
+
+	c.src.bucket = a[0]
+	c.src.key = b[0]
+	if len(b) == 2 {
+		c.src.version = aws.String(b[1])
+	}
+
+	return nil
+}
+
+// getSourceSize returns the size (in bytes) of the source object.
+func (c *copier) getSourceSize() (int64, error) {
+	out, err := c.cfg.S3.HeadObjectWithContext(
+		c.ctx, &s3.HeadObjectInput{
+			Bucket:    &c.src.bucket,
+			Key:       &c.src.key,
+			VersionId: c.src.version,
+		}, c.cfg.RequestOptions...)
+
+	if err != nil {
+		return 0, err
+	}
+
+	return *out.ContentLength, nil
+}
+
+func (c *copier) simpleCopy() (*s3.CopyObjectOutput, error) {
+	return c.cfg.S3.CopyObjectWithContext(c.ctx, c.in, c.cfg.RequestOptions...)
+}
+
+func (c *copier) multipartCopy() (*s3.CopyObjectOutput, error) {
+	upload, err := c.createUpload()
+	if err != nil {
+		return nil, err
+	}
+
+	// set to false once the upload has been completed
+	shouldAbortUpload := true
+	defer func() {
+		if shouldAbortUpload && !c.cfg.LeavePartsOnError {
+			_, _ = c.abortUpload(upload)
+		}
+	}()
+
+	completedParts := make([]*s3.CompletedPart, c.partCount)
+	nextPartNum := int64(0)
+	wg := sync.WaitGroup{}
+	cctx, cancel := canceller(c.ctx)
```

Comment:
Should this be followed with `defer cancel()`? Vet complains of a potential context leak since there are paths where `cancel` is not used.

Reply:
Correct the `cancel` must be called in all paths. The `defer cancel()` is a great way to do that  after creating the cancel context.